openmmtools/tests/test_sampling.py

```diff
@@ -1442,6 +1443,58 @@ def test_online_analysis_stops(self):
             assert sampler._iteration < n_iterations
             assert sampler.is_completed
 
+    def test_context_cache_default(self):
+        """Test default behavior of context cache attributes."""
+        sampler = self.SAMPLER()
+        global_context_cache = cache.global_context_cache
+        # Default is to use global context cache for both context cache attributes
+        assert sampler.sampler_context_cache is global_context_cache
+        assert sampler.energy_context_cache is global_context_cache
+
+    def test_context_cache_energy_propagation(self):
+        """Test specifying different context caches for energy and propagation in a short simulation."""
+        thermodynamic_states, sampler_states, unsampled_states = copy.deepcopy(self.alanine_test)
+        n_replicas = len(sampler_states)
+        if n_replicas == 1:
+            # This test is intended for use with more than one replica
+            return
+
+        with self.temporary_storage_path() as storage_path:
+            # For this test to work, positions should be the same but
+            # translated, so that minimized positions should satisfy
+            # the same condition.
+            original_diffs = [np.average(sampler_states[i].positions - sampler_states[i + 1].positions)
+                              for i in range(n_replicas - 1)]
+            assert not np.allclose(original_diffs,
+                                   [0 for _ in range(n_replicas - 1)]), "sampler %s failed" % self.SAMPLER
+
+            # Create a replica exchange that propagates only 1 femtosecond
+            # per iteration so that positions won't change much.
+            move = mmtools.mcmc.IntegratorMove(openmm.VerletIntegrator(1.0 * unit.femtosecond), n_steps=1)
+            sampler = self.SAMPLER(mcmc_moves=move)
+            reporter = self.REPORTER(storage_path)
+            self.call_sampler_create(sampler, reporter,
+                                     thermodynamic_states, sampler_states,
+                                     unsampled_states)
+            # Set context cache attributes
+            sampler.energy_context_cache = cache.ContextCache(capacity=None, time_to_live=None)
+            sampler.sampler_context_cache = cache.ContextCache(capacity=None, time_to_live=None)
+            # Compute energies
+            sampler._compute_energies()
+            # Check only energy context cache has been accessed
+            assert sampler.energy_context_cache._lru._n_access == sampler.n_replicas
```

Comment:
I'd have the assertion failure print out a more informative error message, such as
```python
            assert sampler.energy_context_cache._lru._n_access == sampler.n_replicas, f'sampler.energy_context_cache._lru._n_access = {sampler.energy_context_cache._lru._n_access} (expected {sampler.n_replicas})'
```
but I think the proper way to test this is just to test for `> 0` (== we are using this cache) vs `==0` (== we are not using this cache). That may be all we need to change.
```

Reply:
I see, that sounds good! And yes, I already made the changes for more informative error. Will push them along with this new way of testing it. Thanks!